### PR TITLE
Reaping the rewards

### PR DESCRIPTION
**Pull request by alexdowad**
It's time to start reaping the rewards of the recent cleanup job.

Now that the code is in so much better shape, it is easier than ever to make functional improvements. 

The first improvement is a new scanner which can handle all, and I mean **ALL**, of Ruby's wild and wooly syntax. Ever wanted to use `%r{}` regexp literals, with nested interpolations (`#{ ... }`) inside the regexps, with multiple `<<HEREDOCS` inside the interpolations, with more deeply nested interpolations inside the `<<HEREDOCS`, right in your very own Racc grammar actions?? No?? Well, if you ever just feel like it, we've got you covered!!!

The next item up on the menu will be fresh and sweet error messages and general diagnostics. Rather than this:

```
10 useless nonterminals
```

I want Racc to do something like clang's delicious diagnostics:

```
statement: tTOKEN1 tTOKEN2 tTOKEN3;
^^^^^^^^^
Useless nonterminal `statement` is not used on the RHS of any other rule, nor is it the start symbol.
```

If `stdout` is going to a TTY, that will be syntax-colored. Coming soon...

(This PR is based on #61, which is based on #58.)

### CODE REVIEW

**Comment by tenderlove** [comment]
@alexdowad I've added you to the collaborators, so feel free to start pushing stuff in there.  I can't find the comment in the flurry of emails but I remember you mentioning emitting a C version of the generated parser.  I think it would also be interesting to experiment with marshalling the state tables to a binary file and then just unmarshalling on boot.  I've noticed Ruby parse time is the bottleneck for loading generated parsers, so maybe just marshalling the state tables would be a win. </randomthoughts>

**Comment by alexdowad** [comment]
@tenderlove Awesome idea, I hadn't thought of that.

Right now I am relying heavily on the byte-for-byte comparison with "known good" outputs to make very sure I don't break anything. Before I try your idea about serializing to some compact binary format, I'll have to come up with another robust testing strategy.

**Comment by alexdowad** [comment]
And I forgot... thank you!

Any chance of a v1.4.14 release now that the problem with error recovery at EOF is fixed?

**Comment by tenderlove** [comment]
Yes, I'll do a .14 release tomorrow. Thanks for fixing it!

**Comment by alexdowad** [comment]
Dear Racc people, while I'm still working on this PR, some review would be appreciated at this point.

I have audited all the CLI options for the `racc` script and found several of them rather useless:

`--line-convert-all` and `--no-line-convert`: controlling whether exceptions raised from within a Racc-generated parser should bear the line numbers from the generated Ruby code, or from the Racc grammar source. I think that showing line numbers from the Racc source is far more useful all around.

`--no-omit-actions`: this tells Racc that for each reduce rule which has no action block, it should generate an empty method and call that method when the reduction is done. Why? For what? It doesn't make any sense.

`-S` versus `--verbose`: `-S` makes Racc print status messages showing its internal "state" from time to time. `--verbose` makes it save a very arcane debugging log file. Why the dichotomy? These have been merged into a single `--verbose` flag. Any "technical" or "internal" information which we might want to log is controlled by this one flag.

`--superclass`: Another head-scratcher. To usefully write custom superclasses for generated parsers, you would need to have quite intimate knowledge of the internals of the generated code. Those internals are not spec'd, not documented, and can validly change at any time. So why provide an option like this? It doesn't make sense.

If someone really wants to monkey-patch a generated parser, they can do this by mixing in a custom module, post-processing the generated source, or patching the parser generator to do what they want.

`-P`: profile Racc itself. This is only useful to the developers and maintainers of Racc, not to users. So why clutter up the CLI interface? An (undocumented) environment variable can be used instead to activate Racc's profiling mode.

Any comments on the above?

The warnings printed when you have conflicts, etc. in your grammar are **far** more informative now. Output is now in Technicolor, though the color scheme is a bit on the hideous side.

Aside from the helpful information printed by default when you have conflicts, `--verbose` adds even more information... which I personally find quite useful, but am not sure if it will make sense to anyone else. Please pull this branch and try `ruby -I./lib bin/racc -v -o /dev/null ...` on a grammar with some conflicts! I'd really like to hear what you think about the diagnostics!

DOT visualization of the LALR state graph is coming next.

**Comment by alexdowad** [comment]
> --superclass: Another head-scratcher.

Hmm. Having said this, I just noticed that the `parser` gem does use `--superclass`. I've looked over the custom superclass, but have yet to figure out whether there is really a good reason for it to be a superclass of the generated parser classes or not.

**Comment by whitequark** [comment]
This is some really impressive work.

**Comment by alexdowad** [comment]
> This is some really impressive work.

Thanks for the encouragement! Please also offer specific opinions on the issues raised above, if you have any!

**Comment by whitequark** [comment]
I'm actually using `--superclass` in Parser, to inject methods common to all parsers. You can just inherit from `Racc::Parser`...

Using dot to indicate token position might be more common but it's harder to notice than underscore (although underscore is not visible too, I personally prefer `^` or even several, spanning entire token).

Line numbers in warnings are useful because they allow me to go straight to the problematic symbol without searching. Editors also have integration that allow to go to `path:line` or `path:line:col` with a single keypress.

Have you considered just cloning clang's color scheme wholesale? clang uses it, gcc uses it, swift uses it (I think), rust uses it (to a degree); it seems just good enough and not surprising to users.

Otherwise LGTM, especially the new diagnostics seem neat. Although it's not like I do much racc hacking anymore...

**Comment by alexdowad** [comment]
> I'm actually using --superclass in Parser, to inject methods common to all parsers. You can just inherit from Racc::Parser...

Hmm. Do you think there is a compelling case to keep this option? Is monkey-patching `Racc::Parser`, or mixing in a custom module, not sufficient for your use case?

I think that before long, `Racc::Parser` may "go away" as a superclass. I would like Racc to compile grammars to code, rather than to tables which drive the `Racc::Parser` automaton. When we get there, there will be little reason for `Racc::Parser` to exist (except to support legacy client code).

> Using dot to indicate token position might be more common but it's harder to notice than underscore (although underscore is not visible too, I personally prefer ^ or even several, spanning entire token).

That means `^` on the following line, right?

> Line numbers in warnings are useful...

OK, let me add these.

> Have you considered just cloning clang's color scheme wholesale?

No, I hadn't. Just trying to figure out what clang's color scheme actually is...

**Comment by whitequark** [comment]
> Is monkey-patching Racc::Parser, or mixing in a custom module, not sufficient for your use case?

What? Why the hell would you even consider suggesting to replace a perfectly local inheritance with monkey-patching a gem I do not control? And `Racc::Parser` going away doesn't make my use case any less compelling, maybe more.

Anyway, it doesn't matter for me, because inheritance chain of parsers is part of public API and I'm pretty sure some code, somewhere, does `.is_a?(Parser::Base)`. So if I cannot keep using that, I will not upgrade to newer racc.

> That means ^ on the following line, right?

I initially mis-remembered how it worked, and now that I remember that it's inline, a dot is fine. So disregard this.

> Just trying to figure out what clang's color scheme actually is...

The basic palette is that notes are dark gray, warnings are violet, errors are red, and squiggly position indicators are green. To see how exactly it looks, try to do `clang -c t.c` containing `int main(;` or something like that.

**Comment by alexdowad** [comment]
> The basic palette is that notes are dark gray, warnings are violet, errors are red, and squiggly position indicators are green.

Ah, OK. That sounds good. But I still need to figure out what nonterminals and terminals should look like.

**Comment by whitequark** [comment]
(See my edit on that comment, and please look at the specific way clang formats errors.)

**Comment by whitequark** [comment]
I think you could underline terminals or something like that.

**Comment by alexdowad** [comment]
Another significant change here:

To manually resolve conflicts, Racc allows you to specify the precedence and associativity of terminals. You can also give a rule an explicit precedence using `=<TOKEN NAME>`.

Here's the thing: if you put explicit precedence on a rule, that precedence *only* applied when deciding whether to **reduce** by that rule or not. It didn't affect the decision of whether to **shift** by that rule. The precedence of all shift operations was always the same as the precedence of the shifted terminal, if in fact the shifted symbol was a terminal.

I have changed things so that when a rule has explicit precedence, the precedence affects the decision to reduce OR to shift by that rule.

With a given state/lookahead token, there can be more than one rule directing the parser to shift. If so, it will take the highest precedence of all of them, and compare that to the precedence of any reduce rule which applies.

If a shift rule has no explicit precedence, it inherits the precedence of the terminal being shifted. If a reduce rule has no explicit precedence, it inherits the precedence of the last terminal in the rule.

Does anyone foresee any problems with the above??

**Comment by whitequark** [comment]
-1 on this change, this isn't how LALR parser generators generally work (in particular not how bison works) and this *will* be unexpected, not to mention the potential to change semantics of existing grammars.

**Comment by alexdowad** [comment]
> -1 on this change, this isn't how LALR parser generators generally work

OK. So how do you manually resolve one specific conflict, without affecting a bunch of other rules at the same time?

**Comment by whitequark** [comment]
You can introduce a different construct for setting precedence for S/R conflicts, just don't change the meaning of existing ones.

**Comment by alexdowad** [comment]
> Anyway, it doesn't matter for me, because inheritance chain of parsers is part of public API and I'm pretty sure some code, somewhere, does .is_a?(Parser::Base). So if I cannot keep using that, I will not upgrade to newer racc.

Fair enough. If `parser` needs it, that is probably a strong enough reason to keep `--superclass`. It would be nice to also get feedback from other interested parties.

**Comment by alexdowad** [comment]
> You can introduce a different construct for setting precedence for S/R conflicts, just don't change the meaning of existing ones.

I just read the docs for `bison`'s precedence declarations. Since `yacc`/`bison` have so much history and such broad adoption in this space, we don't want to adopt a construct from them and change its meaning. You are right about that. (Although they use `%prec`, we use `=`... <shrug>.)

So we may need a new way to explicitly resolve one specific S/R conflict...

Would it be really ugly and verbose to write something like:

```
prefer 'a : T1 T2 T3' over 'a :'
```

??

**Comment by whitequark** [comment]
That looks fine to me.

**Comment by alexdowad** [comment]
`--superclass` has been restored, my ill-considered change to explicit rule precedence has been removed.

**Comment by whitequark** [comment]
Thanks!

**Comment by tenderlove** [comment]
I definitely use --verbose. Never used -S

**Comment by alexdowad** [comment]
> I definitely use --verbose.

What specifically do you use in the `.output` log? Probably the listing of states, right?

**Comment by drbrain** [comment]
That's what I've looked at most

**Comment by alexdowad** [comment]
> That's what I've looked at most

Sorry, what is "that"?

**Comment by alexdowad** [comment]
Oh... sorry. You mean the listing of states. I get it.

**Comment by alexdowad** [comment]
Today we will have DOT visualization of the state graph. Once that hits this branch, can you please try it and see if it makes a good replacement for the listing of states, or whether we still need that listing?

If the conclusion is that we still need the listing of states, that is fine.

**Comment by alexdowad** [comment]
Another question. Does anybody care about `y2racc` and `racc2y`?

There are no tests for these scripts. If there is a compelling use case for them, then tests should be written. If there is no great demand for them, they can be dropped (and anybody who wants them can use the 1.4 gem).

**Comment by drbrain** [comment]
I would drop `y2racc` and `racc2y`, but leave a message in the release notes that if someone wants to resurrect them they're welcome to submit tests, etc.

**Comment by alexdowad** [comment]
Sorry, DOT visualization is still not ready. We do have:

- Colors which don't look like something out of a horror movie
- Racc warns you if you use `=<prec>` to give a rule explicit precedence, but the explicit precedence doesn't help to resolve any conflicts
- Line numbers in warnings
- Rules are printed in warnings just as they are written in the input file, BUT with indentation adjusted to remove excess whitespace while keeping everything aligned as it was in the input, AND with syntax coloring. ALSO, if the "problem" rule is part of a "rule block" (multiple rules separated by `|`), it filters out the lines which are not relevant, while still showing the target nonterminal. Like this:

screenshot from 2015-11-26 16 24 33

To my eyes, this is starting to look really, really sweet. I wish that I could use arbitrary RGB values for colors... but that is a non-standard extension to ANSI which most terminal emulators don't support.

**Comment by alexdowad** [comment]
The code which does the formatting and highlighting you see above is horrendous. It's hairier than a grizzly bear. I hope to clean it up after there is some good test support in place.

**Comment by alexdowad** [comment]
Oh, one more notable change: the codename. Try `./bin/racc --version`.

**Comment by alexdowad** [comment]
- We now warn about nonterminals which are useless because all their rules contain "infinite loops". Racc 1.4 also warns, but the message is much less informative.
- Some error (not warning) messages have been made more informative.
- I found that while showing all the rules which are "active" (or relevant to the current parse) at the points where S/R or R/R conflicts occur was useful, but too verbose. Instead, when warning about a conflict, Racc now shows: 1) a sample of input which would lead to the state with the conflict, and 2) sample inputs which could lead to a successful parse AFTER the conflict, whichever way you go.

In other words, if it's an S/R conflict, it shows you a sample of how a successful parse could be possible after shifting OR after reducing. This is less verbose, and (I find) even more helpful in understanding the reason for the conflict.

**Comment by whitequark** [comment]
:+1: on the last change, Menhir also does that and it's great.

**Comment by alexdowad** [comment]
> Menhir also does that and it's great

First time I ever heard of Menhir -- I'm not an OCaml programmer. Looked at their web page and noted that they already have a couple of other features which I was planning to add to Racc...

**Comment by whitequark** [comment]
I sort of assumed you were copying Menhir features, because so many of the things are so similar.

**Comment by alexdowad** [comment]
"Great minds think alike... and small minds seldom differ."

**Comment by alexdowad** [comment]
Just experimenting with DOT visualization of the state graph. Converted `opal`'s parser states to a DOT file (968 nodes, thousands of arrows). Graphviz is still grinding away on that file, after more than 7 hours.

**Comment by alexdowad** [comment]
Looks like the 16,000+ state transitions of `opal`'s Ruby parser is a bit too much for Graphviz.

What if we restrict the graph to only show the states which are traversed when parsing a given sequence of tokens? 5150c66 is an experiment along that line.

Here is a graph showing the states traversed by `opal`'s parser when parsing `tIDENTIFIER tEQL tINTEGER`. It's very plain; if it is determined that this is a worthwhile feature to add, I will try to "dress it up" to make good use of color, etc:

graph

The graph doesn't show arrows for reduces, so when running through the graph, you have to mentally "step back" past a number of nodes equal to the number of symbols on the RHS of the reduce rule.

I think it would be really cool to produce an animation, using color to indicate the current state in each frame, as the parser consumes input tokens.

To all readers: please be brutally honest if you think that this kind of graph is not useful, and you would prefer a listing of states in some other format.

**Comment by whitequark** [comment]
FYI you can use `-Tsvg` and open it in Chromium. Chromium handles 1GB+ SVG files easily.

**Comment by whitequark** [comment]
But also I don't think this kind of graph is of any use. Take a look at what Menhir does, tbh.

**Comment by alexdowad** [comment]
> Take a look at what Menhir does

Their command-line diagnostics? Or their "interactive interpreter"? Or something else?

**Comment by alexdowad** [comment]
I would like to take this arc of development a bit further, but I'm running out of steam and think it is time to work on something else.

Racc now can generate an HTML summary of parser states which is far more useful than the previous `.output` log. To generate HTML output (rather than an actual parser), use the -S option.

When there is an S/R or R/R conflict, I would like to find a series of tokens which is accepted *both* by an automaton which follows one path, and by an automaton which follows the other. Then I would like to use ASCII art to show 2 different derivation trees for the same series of tokens.

It turns out that this is very difficult in the general case. Actually, not just difficult. It's impossible to solve this problem in the general case, since it reduces to the problem of determining whether 2 context-free grammars are disjoint or not, and that is known to be an undecidable problem. As in, no finite program could ever give the right answer on all possible inputs, *and* always terminate in a finite amount of time.

So that kind of puts a damper on things. But, we don't have to solve this problem in every possible case. Just doing a brute-force search for a solution and bailing out if we don't find one quickly is enough. But I need more powerful tools for manipulating grammars and LR automatons first. Currently, the internals are oriented heavily toward doing what Racc needs to do to generate a parser, not as a general-purpose framework for manipulating and analyzing grammars.
